### PR TITLE
Update STYLE.md with instructions that actually work

### PR DESCRIPTION
**Pull request by ravenblackx**
Signed-off-by: Raven Black <ravenblack@dropbox.com>

Commit Message: Update STYLE.md with instructions that actually work
Additional Description: While trying to send a pull request for an extension I was baffled multiple times by poor error messages and surprising behaviors. The alterations here each derive from a confusing thing that mostly also didn't get any clear "someone understood it better" signs from the dev slack channel.
First surprise, if you use `option (udpa.annotations.file_status).work_in_progress = true;` like the doc suggested, proto_format.sh would remove it. Suggestion from the slack channel was to use `xds.annotations.v3.file_status` instead, which this updates STYLE.md to reflect. ([related slack thread](https://envoyproxy.slack.com/archives/C78HA81DH/p1649780849908359))
Added a link to security posture and status documentation, which was otherwise hard to make sense of without foreknowledge.
Adjusted the section that suggests using `work_in_progress=true` *or* `package_version_status=ACTIVE` - any combination of these (with either xds or udpa version of work_in_progress) that doesn't involve `package_version_status=ACTIVE` causes proto_format.sh to delete the file entirely. So the only viable options appear to be just `package_version_status=ACTIVE`, or that *and* xds `work_in_progress=true` which is already covered in an earlier bulletpoint.
Clarified that without ACTIVE, proto_format.sh will delete the file (which hopefully will make it more discoverable for developers experiencing that symptom).
Added a mention of `api/BUILD`'s `v3_protos` section, without which proto_format.sh will automatically remove any imports of your new proto from other protos.
Added mention of `[#extension:]` tag, which is also necessary, and extra detail about where `[#extension-category:]` has to go (it doesn't appear to work if present at the top-level). ([related slack thread](https://envoyproxy.slack.com/archives/C78HA81DH/p1649799075719279))

Risk Level: None, documentation only.
Testing: Verified that the old documentation doesn't work for at least one developer, and the new documentation is better. :)
Docs Changes: That's all this is.
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: api/STYLE.md [hunk 1]
```diff
@@ -141,12 +141,11 @@ To add an extension config to the API, the steps below should be followed:
...
-1. Make sure your proto is either tracked as a work in progress
-   (`option (udpa.annotations.file_status).work_in_progress = true;`)
-   or ready to be used
+1. Make sure your proto is tracked as ready to be used
    (`option (udpa.annotations.file_status).package_version_status = ACTIVE;`).
-   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD).
-1. Add a reference to the v3 extension config in (1) in [api/versioning/BUILD](versioning/BUILD) under `active_protos`.
+   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD), and without it `proto_format.sh` will delete the proto file.
+1. Add a reference to the v3 extension config in (2) in [api/versioning/BUILD](versioning/BUILD) under `active_protos`.
+1. Add a reference to the v3 extension config in (2) in [api/BUILD](BUILD) under `v3_protos`. Without this, `proto_format.sh` will erase any imports of your proto file from other proto files.
```
Comment (mattklein123) [inline api/STYLE.md:148]: Neither of these sound right since both of these files are auto generated, but I haven't looked at this in a while. @phlax @htuch @adisuissa can you advise? @ravenblackx I'm wondering if the problem you are having is that you are not actually consuming your new protos in an actual extension yet, you are just doing a prep/common PR.

**Review by mattklein123 — COMMENTED** [review]
Really sorry for the trouble. We have been trying to make this tooling simpler but it's still pretty confusing. I've CCd a few people to help look at this.

File: api/STYLE.md (hunk 1 above)
Comment (ravenblackx) [reply api/STYLE.md:148]: I agree that the active_protos one seems wrong (that bullet was already in the doc, and I was torn as to whether to remove it!) I had observed that one being auto-generated when I run proto_format.sh.

The v3_protos one doesn't seem to be getting auto-generated, and it's not that it's unused (the 'testing' was in PR20332 which is at least *somewhat* using it in an extension). The other proto file that was trying to import it and getting reverted by proto_format.sh is a prep PR that doesn't use *that* proto yet, but it was the one from PR20332 that lacked the v3_protos entry, which was the cause of that problem.

Comment (htuch) [reply api/STYLE.md]: First, I agree this is confusing. What's going on is a form of fixed function computation. Despite the warning on the file and the fact it is automatically generated, it does in fact rely on an initial addition of a new file to allow the tool to converge to the new state. So, the advice of updating `api/BUILD` is correct, we should change the warning at the top of the file to reflect the use here. We could also possibly simplify this tooling significantly further or make its data flow / dependency more explicit.

For the curious, what is going on is that `api/BUILD` is auto-generated, as per https://github.com/envoyproxy/envoy/blob/main/api/BUILD#L1. It is generated by `tools/type_whisperer/proto_build_targets_gen.py`. This relies on the type database as source of truth, https://github.com/envoyproxy/envoy/blob/ba362a22b43a41379b156563f2f83402762081bf/tools/type_whisperer/BUILD#L64. The type database is built from `api/BUILD` (https://github.com/envoyproxy/envoy/blob/ba362a22b43a41379b156563f2f83402762081bf/tools/type_whisperer/BUILD#L59).

Comment (adisuissa) [reply api/STYLE.md]: I agree that the simple way forward is clarifying the comment here:
https://github.com/envoyproxy/envoy/blob/2e9179f6ab7fe044fc3d9419452251101e4b6cc8/tools/type_whisperer/proto_build_targets_gen.py#L39

An alternative approach could be traversing the proto files, and creating a list of v2_deps and v3_deps.

File: api/STYLE.md [hunk 2]
```diff
@@ -130,9 +130,9 @@ To add an extension config to the API, the steps below should be followed:
    )
    ```
 1. Update [source/extensions/extensions_metadata.yaml](../source/extensions/extensions_metadata.yaml)
-   with the category, security posture, and status. The category field will have to match an
+   with the category, [security posture, and status](../EXTENSION_POLICY.md#extension-stability-and-security-posture). The category field will have to match an
    annotation of the form `// [#extension-category: your.extension.category]`
-   in one of the proto files for the docs build to pass.
+   in a message in one of the proto files for the docs build to pass. You may also need `// [#extension: your.extension.category.extension]`.
```
Comment (adisuissa) [inline api/STYLE.md:135]: nit:
```suggestion
   in a message in one of the proto files for the docs build to pass. You may also need `// [#extension: your.extension.category.extension_name]`.
```

File: api/STYLE.md [hunk 3]
```diff
@@ -189,7 +187,7 @@ metadata. We describe these annotations below by category.
 * `option (udpa.annotations.file_migrate).move_to_package = "<package name>";`
   to denote that in the next major version of the API, the file will be moved to
   the given package. This is consumed by `protoxform`.
-* `option (udpa.annotations.file_status).work_in_progress = true;` to denote a
+* `option (xds.annotations.v3.file_status).work_in_progress = true;` to denote a
```
Comment (adisuissa) [inline api/STYLE.md:190]: @htuch any preference between `xds.annotations...` and `udpa.annotations...`? (not only for WIP but also other annotations)

File: api/STYLE.md [hunk 4]
```diff
@@ -141,12 +141,10 @@ To add an extension config to the API, the steps below should be followed:
...
 1. Make sure your proto imports the v3 extension config proto (`import "udpa/annotations/status.proto";`)
-1. Make sure your proto is either tracked as a work in progress
-   (`option (udpa.annotations.file_status).work_in_progress = true;`)
-   or ready to be used
+1. Make sure your proto is tracked as ready to be used
    (`option (udpa.annotations.file_status).package_version_status = ACTIVE;`).
-   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD).
-1. Add a reference to the v3 extension config in (1) in [api/versioning/BUILD](versioning/BUILD) under `active_protos`.
+   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD) under `active_protos`, and without it `proto_format.sh` will delete the proto file.
+1. Add a reference to the v3 extension config in (2) in [api/BUILD](BUILD) under `v3_protos`. Without this, `proto_format.sh` will erase any imports of your proto file from other proto files.
```
Comment (adisuissa) [inline api/STYLE.md:147]: I agree with what you are writing, just trying to get some clarity on why the line was erased.
IIRC this is what happened, but feel free to correct me.

You introduced 2 new protos, lets call them A and B, where B imported A.
When you ran `proto_format` it removed the import line from B.
Was the line deleted because the A wasn't in the BUILD file or was it because B wasn't there (or both).
I'm assuming that it was because A wasn't declared in the `api/BUILD`.

If one would first add extension A, run the script, and only then add B and run the script, would it still erase the import line in B?

**Review by adisuissa — COMMENTED** [review]
Thanks for clarifying the process!

File: api/STYLE.md (hunk 1 above)
Comment (ravenblackx) [reply api/STYLE.md]: Agree that it would be best for the tooling to *actually* do this automatically, and second-best to update the DO NOT EDIT in addition to STYLE.md.
As a short-term fix to just make it *less* confusing until either one of those things happens, the proposed additional bullet in STYLE.md is sufficient to at least make it a 5 minute confusion instead of a 2 hour confusion. I'd rather not expand this change out into also touching .py files.

File: api/STYLE.md (hunk 4 above)
Comment (ravenblackx) [reply api/STYLE.md:147]: Actual sequence:
I added proto A, and ran `proto_format`. `api/BUILD` remains untouched.
I added proto B, importing A, and ran `proto_format` again. The import of A was removed.
I undid that change.
I added proto A to api/BUILD, and ran `proto_format` again. The import of A is not removed.

Nothing actually goes wrong (I believe CI even passes) if you don't add proto A to `api/BUILD`, until another proto tries to depend on it (that was even in a separate PR in this case). `api/BUILD` doesn't get automatically populated at all, because the thing that *would* automatically populate it doesn't see the thing to add unless you already added it. So I think it's essentially just acting as an auto-formatter of that build file, not an auto-populater. (Similar to how you have to add the proto's build dependencies to the "do not edit" BUILD file before the script will run and "auto-generate" that BUILD file - but worse because for *that* case you get an error message, vs. for `api/BUILD` there's not even a warning that stuff wasn't generated).

**Comment by ravenblackx** [comment]
Sorry about force-push - it was required because apparently clicking to apply a suggestion commit makes an unsigned commit which upsets DCO...

File: api/STYLE.md (hunk 4 above)
Comment (adisuissa) [reply api/STYLE.md:147]: Thanks.
For some reason I was under the impression the `proto_format` will add the proto to the `api/BUILD` file, but if it doesn't, then the comment in `api/BUILD` is wrong (not only needs to be clarified).

File: api/STYLE.md (hunk 3 above)
Comment (htuch) [reply api/STYLE.md:190]: We're in the process of moving all `udpa.*` to `xds.*`, so the former is better.

File: api/STYLE.md (hunk 4 above)
Comment (ravenblackx) [reply api/STYLE.md:147]: The comment in every api/**/v3/BUILD file is wrong in the same way. They all say `# DO NOT EDIT. This file is generated by tools/proto_format/proto_sync.py.`, but you have to edit them manually if you want to add an import or proto_format.sh won't run.

I definitely would rather not include that change into this PR, since changing the generator output will also require updating every generated-ish file it touches.

Comment (htuch) [reply api/STYLE.md:147]: I think it's not really wrong for `api/**v3/BUILD`. You have to populate this file once, initially, when you create a new package. After that point it's automatically maintained and not changed. I think we should be discouraging editing after the initial checkin in most of these files (that's the point of the warning).

File: api/STYLE.md [hunk 5]
```diff
@@ -116,7 +116,7 @@ organization](#package-organization) above.
 To add an extension config to the API, the steps below should be followed:
 
 1. If this is still WiP and subject to breaking changes, please tag it
-   `option (udpa.annotations.file_status).work_in_progress = true;` and
+   `option (xds.annotations.v3.file_status).work_in_progress = true;` and
    optionally hide it from the docs (`[#not-implemented-hide:]`.
```
Comment (phlax) [inline api/STYLE.md:120]: nit: not a change in this PR - but this sentence is missing a closing parenthesis

Comment (ravenblackx) [reply api/STYLE.md]: Added it.

File: api/STYLE.md [hunk 6]
```diff
@@ -130,9 +130,9 @@ To add an extension config to the API, the steps below should be followed:
    )
    ```
 1. Update [source/extensions/extensions_metadata.yaml](../source/extensions/extensions_metadata.yaml)
-   with the category, security posture, and status. The category field will have to match an
+   with the category, [security posture, and status](../EXTENSION_POLICY.md#extension-stability-and-security-posture). The category field will have to match an
    annotation of the form `// [#extension-category: your.extension.category]`
-   in one of the proto files for the docs build to pass.
+   in a message in one of the proto files for the docs build to pass. You may also need `// [#extension: your.extension.category.extension_name]`.
```
Comment (phlax) [inline api/STYLE.md:135]: this is a bit wordy and hard to follow 

are we saying something like...

> any extension category added to the metadata.yaml file should be present in one and only one proto file as an annotation (...)

and

> every configured extension category should have at least one extension defined. This is can be done in the relevant proto file with the annotation (...)


not sure if my suggestion is helpful or correct, but i think these sentences need breaking up and some clarification

File: api/STYLE.md [hunk 7]
```diff
@@ -141,12 +141,10 @@ To add an extension config to the API, the steps below should be followed:
...
    See the [key-value-store PR](https://github.com/envoyproxy/envoy/pull/17745/files) for an example of adding a new extension point to common.
 1. Make sure your proto imports the v3 extension config proto (`import "udpa/annotations/status.proto";`)
-1. Make sure your proto is either tracked as a work in progress
-   (`option (udpa.annotations.file_status).work_in_progress = true;`)
-   or ready to be used
+1. Make sure your proto is tracked as ready to be used
    (`option (udpa.annotations.file_status).package_version_status = ACTIVE;`).
-   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD).
-1. Add a reference to the v3 extension config in (1) in [api/versioning/BUILD](versioning/BUILD) under `active_protos`.
+   This is required to automatically include the config proto in [api/versioning/BUILD](versioning/BUILD) under `active_protos`, and without it `proto_format.sh` will delete the proto file.
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply api/STYLE.md]: Done.

File: api/STYLE.md (hunk 6 above)
Comment (ravenblackx) [reply api/STYLE.md]: I've had a go at making this clearer (and in the process also clarified my own understanding of what these annotations mean!)
I'm not fully sure why extension-category must be in only one place, it seems like theoretically it should be possible to have more than one thing extended by the same category of extension, but I didn't find an example of that happening so I guess it might be an arbitrary constraint.

Now documented with codeblock examples so you can easily see where both annotations belong.

**Comment by htuch** [comment]
This is great @phlax. Is there anyone in @envoyproxy/first-pass-reviewers who hasn't already written an extension with proto and would have some cycles to run through these instructions? I feel this is the ultimate test :)

**Comment by phlax** [comment]
just to clarify @htuch - @ravenblackx is the author - my only contrib is a minimal review - but agree this is a great improvement

as a way forward im wondering if its best to land and then post to the slack/dev channel to encourage some acceptance testing

**Comment by alyssawilk** [comment]
+1 to that - given this has been stalled with no volunteers for a few

**Review by adisuissa — COMMENTED** [review]
Small nit, but otherwise LGTM. Thanks!
I agree we should probably land this, and fix this as more people test this out.

**Comment by ravenblackx** [comment]
Still found some confusion following the updated doc, it turns out - adjusted it to be more helpful in those regards.

(I had the `#extension` annotation above the third message in the proto file, which is the actual config message - doing that causes the annotation to not be usefully detected, and you don't get a helpful error like "this annotation should be at the top", it just acts the same as if the annotation was absent from the file entirely.)

**Review by htuch — APPROVED** [review]
LGTM, thanks!
